identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccess.kt

```diff
@@ -0,0 +1,493 @@
+        val scratchpad = ByteArray(2)
+        val bos = ByteArrayOutputStream()
+        // set instruction
+        setShort(scratchpad, 0, cmd.toShort())
+        bos.write(scratchpad)
+        bos.write(slot)
+        bos.write(operation.toInt())
+        bos.write(Cbor.encode((Bstr(bb.array()))))
+        val beginApdu = makeCommandApdu(bos.toByteArray())
+
+        val response: ByteArray = transport.sendData(beginApdu)
+        val status = getAPDUResponseStatus(response)
+        check(APDU_RESPONSE_STATUS_OK == status) {
+            "Operation failed. Response status: $status" }
+        if (response.size > 2) {
+            val input = response.copyOf(response.size - 2)
+            return decode(input).asBstr
+        }
+        return null
+    }
+
+    /**
+     * Initializes the transport so that the app can begin connecting to the applet in the background.
+     */
+    fun warmupTransport() {
+        DirectAccessOmapiTransport()
+    }
+
+    /**
+     * Allocates space (a slot) in the Direct Access applet for a
+     * document to be provisioned to.
+     *
+     * @return ID of the allocated slot, -1 if no slot is available
+     */
+    fun allocateDocumentSlot(): Int {
+        val apdu: ByteArray?
+        val response: ByteArray?
+        try {
+            transport.closeConnection()
+            transport.openConnection()
+
+            val scratchpad = ByteArray(2)
+            val bos = ByteArrayOutputStream()
+            // set instruction
+            setShort(scratchpad, 0, CMD_MDOC_CREATE.toShort())
+            bos.write(scratchpad)
+            apdu = makeCommandApdu(bos.toByteArray())
+
+            response = transport.sendData(apdu)
+            check(getAPDUResponseStatus(response) == APDU_RESPONSE_STATUS_OK)
+            transport.closeConnection()
+            return response[0].toInt()
+        } catch (e: IOException) {
+            transport.closeConnection()
+            throw java.lang.IllegalStateException("Failed to send createCredential APDU command")
+        }
+    }
+
+    /**
+     * Clears a slot.
+     *
+     * @param documentSlot the slot to delete any existing document at
+     * @return true if the slot is cleared, false otherwise
+     */
+    fun clearDocumentSlot(
+        documentSlot: Int
+    ): Boolean {
+        try {
+            transport.closeConnection()
+            transport.openConnection()
+
+            val scratchpad = ByteArray(2)
+            val bos = ByteArrayOutputStream()
+            setShort(scratchpad, 0, CMD_MDOC_DELETE_CREDENTIAL.toShort())
+            bos.write(scratchpad)
+            bos.write(documentSlot)
+            val apdu = makeCommandApdu(bos.toByteArray())
+
+            val response: ByteArray = transport.sendData(apdu)
+            val status = getAPDUResponseStatus(response)
+            check(APDU_RESPONSE_STATUS_OK == status) {
+                "clearDocumentSlot failed. Response status: $status" }
+            transport.closeConnection()
```

Comment:
This won't be called if the status is not OK. Maybe move it to a finally {} block, since it looks like we always want to do this?